src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs

```diff
@@ -0,0 +1,362 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Intrinsics.X86;
+using System;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    public static partial class CpuMathUtils
+    {
+        public const int CbAlign = 16;
+
+        private static bool Compat(AlignedArray a)
+        {
+            Contracts.AssertValue(a);
+            Contracts.Assert(a.Size > 0);
+            return a.CbAlign == CbAlign;
+        }
+
+        private unsafe static float* Ptr(AlignedArray a, float* p)
+        {
+            Contracts.AssertValue(a);
+            float* q = p + a.GetBase((long)p);
+            Contracts.Assert(((long)q & (CbAlign - 1)) == 0);
+            return q;
+        }
+
+        public static void Scale(float a, float[] dst, int count)
+        {
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 < count && count <= dst.Length);
+
+            if (Sse.IsSupported)
+            {
+                SseIntrinsics.ScaleU(a, dst, count);
+            }
+            else
+            {
+                for (int i = 0; i < dst.Length; i++)
+                {
+                    dst[i] *= a;
+                }
+            }
+        }
+
+        public static void Scale(float a, float[] dst, int offset, int count)
+        {
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 < count);
+            Contracts.Assert(0 <= offset && offset < dst.Length - count);
+
+            ArraySegment<float> dstSeg = new ArraySegment<float>(dst, offset, count);
+            Scale(a, dstSeg.Array, count);
```

Comment:
I'm not sure if this usage of `ArraySegment` is correct.

dstSeg.Array is just going to return the original array `dst`, so we are missing using the offset at all.

We should invert the way we call these two methods instead.  `Scale(float a, float[] dst, int count)` should call `Scale(float a, float[] dst, int offset, int count)` with `offset == 0`.

Reply:
Updating these, long term, to use `Span<T>` would probably be better as well.

Note: Some of my comments, like this, can be logged as "future" work items. Having the initial PR be a "clean" port of the existing code is desirable. Some amount of refactoring or general improvements can happen in a later PR.

Reply:
Nice catch.  I agree that the usage of `ArraySegment` does not achieve the intended result, and that we should have  `Scale(float a, float[] dst, int offset, int count)` call `Scale(float a, float[] dst, int count)` with `offset == 0`.  

The latter means that we should have the Scale method with the offset call an SSE intrinsic, e.g. ScaleU.  However, currently, there is no implementation of SSE intrinsics that accepts an offset.  The native implementation works well with the offset ([here](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.CpuMath/Sse.cs#L635)) because we can increment the input pointer with the offset in C, but in the managed implementation of SSE intrinsics, we need to pass in an array that starts with the `offset`-th element.  Since the Scale functions intends to modify the input arrays in place, making any copy of the original input arrays may not work.  One suggestion is to create new SSE intrinsics in `SseIntrinsics.cs` that accept an offset, and have the original SSE intrinsics call those new intrinsics with `offset == 0`.  This suggested approach will make `SseIntrinsics.cs` no longer completely in sync with `src\Native\Sse.cpp`, but could solve the problem.

Please let me know how the suggested approach sounds.  If it sounds ok, I will modify in this fashion right away.

Reply:
The "proper" way to fix this should be to just use `Span<T>` (IMO). `Span<T>` allows you to do exactly as you suggest and treat `array[offset]` as if it was index 0.

Outside of that, it should be trivial to update `ScaleU` to take an offset (should just be changing `float* pSrcCurrent = psrc` to `float* pSrcCurrent = psrc + offset` and the same for pdst`)

Reply:
Thank you so much! Edited :)